File: frontend/src/types/goal.ts
export type GoalType = 'ultimate' | 'monthly' | 'weekly';
export type GoalStatus = 'active' | 'completed' | 'abandoned';

export interface Goal {
  id: number;
  type: GoalType;
  title: string;
  description: string;
  category_id: number | null;
  parent_goal_id: number | null;
  target_date: string | null;
  status: GoalStatus;
  progress: number;
  completed_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface GoalCreate {
  type: GoalType;
  title: string;
  description?: string;
  category_id?: number | null;
  parent_goal_id?: number | null;
  target_date?: string | null;
}

export interface GoalUpdate {
  title?: string;
  description?: string;
  status?: GoalStatus;
  progress?: number;
  target_date?: string | null;
}

File: backend/app/schemas/goal.py
from datetime import date, datetime
from typing import Literal, Optional

from pydantic import BaseModel, ConfigDict


class GoalCreate(BaseModel):
    type: Literal["ultimate", "monthly", "weekly"]
    title: str
    description: str | None = None
    category_id: int | None = None
    parent_goal_id: int | None = None
    target_date: date | None = None


class GoalUpdate(BaseModel):
    title: str | None = None
    description: str | None = None
    category_id: int | None = None
    target_date: date | None = None


class GoalStatusUpdate(BaseModel):
    status: Literal["active", "completed", "abandoned"]


class GoalResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    type: str
    title: str
    description: str | None
    category_id: int | None
    parent_goal_id: int | None
    target_date: date | None
    status: str
    progress: int
    completed_at: datetime | None
    created_at: datetime
    updated_at: datetime


class GoalProgressResponse(BaseModel):
    goal: GoalResponse
    children: list[GoalResponse]
    related_tasks_count: int
    completed_tasks_count: int
